packages/salesforcedx-vscode-core/src/commands/forceSourceDeploy.ts

```diff
@@ -5,37 +5,129 @@
  * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
 
-import {
-  Command,
-  SfdxCommandBuilder
-} from '@salesforce/salesforcedx-utils-vscode/out/src/cli';
+import { CliCommandExecutor, Command, SfdxCommandBuilder } from '@salesforce/salesforcedx-utils-vscode/out/src/cli';
+import { ContinueResponse } from '@salesforce/salesforcedx-utils-vscode/out/src/types';
+import * as path from 'path';
+import { Observable } from 'rxjs/Observable';
+import * as vscode from 'vscode';
+import { channelService } from '../channels';
 import { nls } from '../messages';
-import {
-  SfdxCommandlet,
-  SfdxCommandletExecutor,
-  SfdxWorkspaceChecker
-} from './commands';
-import {
-  FileType,
-  ManifestOrSourcePathGatherer,
-  SelectedPath
-} from './forceSourceRetrieve';
-
-export class ForceSourceDeployExecutor extends SfdxCommandletExecutor<
-  SelectedPath
-> {
+import { notificationService, ProgressNotification } from '../notifications';
+import { taskViewService } from '../statuses';
+import { SfdxCommandlet, SfdxCommandletExecutor, SfdxWorkspaceChecker } from './commands';
+import { FileType, ManifestOrSourcePathGatherer, SelectedPath } from './forceSourceRetrieve';
+
+interface Result {
+  columnNumber: string;
+  error: string;
+  filePath: string;
+  fullName: string;
+  lineNumber: string;
+  type: string;
+}
+
+interface ForceSourceDeployErrorResult {
+  message: string;
+  name: string;
+  result: Result[];
+  stack: string;
+  status: number;
+  warnings: any[];
+}
+
+// move this elsewhere?
+vscode.workspace.onDidChangeTextDocument((e => {
+  if (ForceSourceDeployExecutor._errorCollection.has(e.document.uri)) {
+    ForceSourceDeployExecutor._errorCollection.delete(e.document.uri);
+  }
+}));
+
+export class ForceSourceDeployExecutor extends SfdxCommandletExecutor<SelectedPath> {
+
+  public static _errorCollection = vscode.languages.createDiagnosticCollection('deploy-errors');
+
   public build(data: SelectedPath): Command {
     const commandBuilder = new SfdxCommandBuilder()
       .withDescription(nls.localize('force_source_deploy_text'))
       .withArg('force:source:deploy')
-      .withLogName('force_source_deploy');
+      .withLogName('force_source_deploy')
+      .withJson();
     if (data.type === FileType.Manifest) {
       commandBuilder.withFlag('--manifest', data.filePath);
     } else {
       commandBuilder.withFlag('--sourcepath', data.filePath);
     }
     return commandBuilder.build();
   }
+
+  public execute(response: ContinueResponse<SelectedPath>): void {
+    const cancellationTokenSource = new vscode.CancellationTokenSource();
+    const cancellationToken = cancellationTokenSource.token;
+
+    const execution = new CliCommandExecutor(this.build(response.data), {
+      cwd: vscode.workspace.rootPath
```

Comment:
`workspace.rootPath` is deprecated (https://github.com/Microsoft/vscode/wiki/Extension-Authoring:-Adopting-Multi-Root-Workspace-APIs), you can use `workspace.workspaceFolders[0]` for now since we need to make further changes for the extensions to support multiple workspaces.